src/types/list.rs

```diff
@@ -465,9 +465,9 @@ impl<'py> BoundListIterator<'py> {
     }
 
     unsafe fn get_item(&self, index: usize) -> Bound<'py, PyAny> {
-        #[cfg(any(Py_LIMITED_API, PyPy))]
+        #[cfg(any(Py_LIMITED_API, PyPy, Py_GIL_DISABLED))]
         let item = self.list.get_item(index).expect("list.get failed");
-        #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+        #[cfg(not(any(Py_LIMITED_API, PyPy, Py_GIL_DISABLED)))]
```

Comment:
I think there's probably also a question of concurrent additions (or removals) to the list during iteration; at the moment we can rely on re-checking the length on each call to `.next()` and the guarantee of the GIL to stop us doing an out-of-bounds read. With the freethreaded build, I assume it's possible to have a time-of-check to time-of-use error between length and `.get_item()` call here.

It seems possible that we can have panics on the freethreaded build from the `"list.get failed"` where they should be rarer (impossible?) on the GIL build.

I wonder, what happens on the freethreaded build if a list is modified (in another thread) during iteration in pure-python?

Should we use `list.tp_iter` rather than defining our own iterator? (i.e. are we back to similar questions as in #4439?)

Reply:
> I wonder, what happens on the freethreaded build if a list is modified (in another thread) during iteration in pure-python?

Then you get a race condition. Try yourself with this script:

```python
import threading

lis = [_ for _ in range(3)]
b = threading.Barrier(2)

def func1():
    b.wait()
    lis[0] = 7

def func2():
    waited = False
    for i in lis:
        if not waited:
            waited = True
            b.wait()
	print(lis)

threads = [threading.Thread(target=func1), threading.Thread(target=func2)]

[t.start() for t in threads]
[t.join() for t in threads]
```

There's also a race with the GIL, but I don't think you'll ever see this script print out

```
[7, 1, 2, 3]
[7, 1, 2, 3]
[7, 1, 2, 3]
[7, 1, 2, 3]
```

with the GIL disabled. With the GIL enabled you don't know when the next time the other thread will be able to run, so you might see

```
[0, 1, 2, 3]
[7, 1, 2, 3]
[7, 1, 2, 3]
[7, 1, 2, 3]
```

but also you might see


```
[0, 1, 2, 3]
[0, 1, 2, 3]
[0, 1, 2, 3]
[0, 1, 2, 3]
```

The same is also true about getting random results on both builds (but not crashing) if you replace the list.setitem with e.g. `lis.append(4)`. So I think it definitely is possible for the list size to change "underneath" a thread, on both builds.

Should we do anything special to account for that possibility?

Reply:
Following up, see https://github.com/PyO3/pyo3/issues/4571#issue-2539376644 for the way forward, if someone really wants a locked iterator they probably should explicitly opt into that. I'll also take a look at implementing `Iterator::fold` and `Iterator::try_fold` today and add those if that turns out to be straightforward.